Builder/jni/djvu/tools/djvups.1
.\" Copyright (c) 2003 Leon Bottou, Yann Le Cun, Patrick Haffner,
.\" Copyright (c) 2001 AT&T Corp., and Lizardtech, Inc.
.\"
.\" This is free documentation; you can redistribute it and/or
.\" modify it under the terms of the GNU General Public License as
.\" published by the Free Software Foundation; either version 2 of
.\" the License, or (at your option) any later version.
.\"
.\" The GNU General Public License's references to "object code"
.\" and "executables" are to be interpreted as the output of any
.\" document formatting or typesetting system, including
.\" intermediate and printed output.
.\"
.\" This manual is distributed in the hope that it will be useful,
.\" but WITHOUT ANY WARRANTY; without even the implied warranty of
.\" MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
.\" GNU General Public License for more details.
.\"
.\" You should have received a copy of the GNU General Public
.\" License along with this manual. Otherwise check the web site
.\" of the Free Software Foundation at http://www.fsf.org.
.TH DJVUPS 1 "01/18/2003" "DjVuLibre-3.5" "DjVuLibre-3.5"
.SH NAME
djvups \- Convert DjVu documents to PostScript.

.SH SYNOPSIS
.BI "djvups  [" "options" "] [" "djvufile" "] [" "outputfile" "]"

.SH DESCRIPTION         
This program decodes DjVu file 
.IR djvufile ,  
and generates a PostScript file named
.IR outputfile .
The DjVu data is read from the standard input when argument 
.IR "djvufile"
is not specified or when it is equal to a single dash.
Similarly, the output data is written to the standard output
when argument 
.IR "outputfile"
is not specified or equal to a single dash.

PostScript printers have various capabilities.
Investigate options 
.B -level
and
.B -gray
for obtaining the best results.

.SH OPTIONS
.TP
.BI -help
Prints the list of recognized options.
.TP
.BI -verbose
Displays a progress bar.
.TP
.BI -page= pagespec
Specify the document pages to be converted.
The page specification
.I pagespec 
contains one or more comma-separated page ranges.
A page range is either a page number, 
or two page numbers separated by a dash.
Specification
.BR 1-10 ,
for instance, prints pages 1 to 10.
Specification
.BR 1,3,99999-4
prints pages 1 and 3, followed by all the document
pages in reverse order up to page 4.
.TP
.BI -format=ps
Produce a PostScript file. 
This is the default.
.TP
.BI -format=eps
Produce an Encapsulated PostScript file.
Encapsulated PostScript files are suitable for
embedding images into other documents.
Encapsulated PostScript file can only contain 
a single page. Setting this option
overrides the options
.BR -copies ,
.BR -orientation ,
.BR -zoom ,
.BR -cropmarks ,
and
.BR -booklet .
.TP
.BI -copies= n
Specify the number of copies to print.
.TP
.BI -orientation= orient
Specify whether pages should be printed using
the
.BR auto ,
.BR portrait ,
or
.B landscape 
orientation.
.TP
.BI -mode= modespec
Specify how pages should be decoded.
The default mode,
.BR color ,
renders all the layers of the DjVu documents.
Mode 
.BR black
only renders the foreground layer mask.  This mode does not work with
DjVuPhoto images because these files have no foreground layer
mask.
Modes
.BR foreground 
and
.BR background
only render the foreground layer or the background layer 
of a DjVuDocument image.
.TP
.BI -zoom= zoomspec
Specify a zoom factor
.IR zoomspec .
The default zoom factor, 
.BR auto ,
scales the image to fit the page.
Argument
.I zoomspec
also can be a number in range 
.BR 25
to
.BR 2400 
representing a magnification percentage
relative to the original size of the document.
.TP
.BI -frame= yesno
Specifying 
.BR yes
causes the generation of a thin gray border
representing the boundaries of the document pages.
The default is
.BR no .
.TP
.BI -cropmarks= yesno
Specifying 
.BR yes
causes the generation of crop marks
indicating where pages should be cut.
The default is
.BR no .
.TP
.BI -level= languagelevel
Select the language level of the generated PostScript.
.IR languagelevel . 
Valid language levels are 
.BR 1 ,
.BR 2 ,
and
.BR 3 .
Level 
.B 3 
produces the most compact and fast printing PostScript files.
Some of these files however require a very modern printer.
Level
.B 2 
is the default value.
The generated PostScript files are almost as compact
and work with all but the oldest PostScript printers.
Level
.B 1
can be used as a last resort option.
.TP
.BI -color= yesno
The default value,
.BR yes ,
generates a color PostScript file.
Specifying value
.BR no
converts the image to gray scale.
The resulting PostScript file is smaller
and marginally more portable.
.TP
.BI -gray
This option is equivalent to option
.BR -color=no 
and is provided for convenience.
.TP
.BI -colormatch= yesno
The default value,
.BR yes ,
generates a PostScript file using
device independent colors in compliance
with the 
.SM sRGB
specification.
Modern printers then produce colors that match
the original as well as possible.  
Specifying value
.B no
generates a PostScript file using device dependent colors.  
This is sometimes useful with older printers.
You can then use option
.BR -gamma 
to tune the output colors.
.TP
.BI -gamma= gammaspec
Specify a gamma correction factor for the device dependent PostScript colors.
Argument 
.I gammaspec
must be in range
.B 0.3
to
.BR 5.0 .
Gamma correction normally pertains to cathodic screens only.  
It gets meaningful for printers because several models
interpret device dependent RGB colors by emulating the color
response of a cathodic tube.  
.TP
.BI -booklet= opt
Turns the booklet printing mode on.
The booklet mode prints two pages on each side
in a way suitable for making a booklet by folding
the sheets. Option
.I opt
can take values
.B no
for disabling the booklet mode,
.B yes
for enabling the recto/verso booklet mode, and
.B recto
or
.B verso
to print only one side of each sheet.
.TP
.BI -bookletmax= max
Specifies the maximal number of pages per booklet.
A single printout might then be composed of several 
booklets. Argument
.I max
is rounded up to the next multiple of 4. 
Specifying 
.B 0
sets no maximal number of pages
and ensures that the printout will produce
a single booklet.  This is the default.
.TP
.BI -bookletalign= align
Specifies a positive or negative offset 
applied  to the verso of each sheet.  
Argument 
.I align
is expressed in points 
(one point is 1/72th of an inch, or 0.352 millimeter)
This is useful with certain printers to ensure that 
both recto and verso are properly aligned.  The default
value is of course
.BR 0 .
.TP
.BI -bookletfold= base[+incr]
Specifies the extra margin left  between both pages
on a single sheet.  The base value
.I base
is expressed in points (one point is 1/72th of an inch, 
or 0.352 millimeter). This margin is incremented for 
each outer sheet by value
.I incr
expressed in millipoints. The default value is 
.BR 18+200 .

.SH CREDITS
This program was written by L\('eon Bottou <leonb@users.sourceforge.net>,
Andrei Erofeev <andrew_erofeev@yahoo.com>, and Florin Nicsa.

.SH SEE ALSO
.BR djvu (1),
.BR ddjvu (1).
.BR djview (1)


Builder/jni/djvu/tools/djvups.cpp
//C-  -*- C++ -*-
//C- -------------------------------------------------------------------
//C- DjVuLibre-3.5
//C- Copyright (c) 2002  Leon Bottou and Yann Le Cun.
//C- Copyright (c) 2001  AT&T
//C-
//C- This software is subject to, and may be distributed under, the
//C- GNU General Public License, either Version 2 of the license,
//C- or (at your option) any later version. The license should have
//C- accompanied the software or you may obtain a copy of the license
//C- from the Free Software Foundation at http://www.fsf.org .
//C-
//C- This program is distributed in the hope that it will be useful,
//C- but WITHOUT ANY WARRANTY; without even the implied warranty of
//C- MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
//C- GNU General Public License for more details.
//C- 
//C- DjVuLibre-3.5 is derived from the DjVu(r) Reference Library from
//C- Lizardtech Software.  Lizardtech Software has authorized us to
//C- replace the original DjVu(r) Reference Library notice by the following
//C- text (see doc/lizard2002.djvu and doc/lizardtech2007.djvu):
//C-
//C-  ------------------------------------------------------------------
//C- | DjVu (r) Reference Library (v. 3.5)
//C- | Copyright (c) 1999-2001 LizardTech, Inc. All Rights Reserved.
//C- | The DjVu Reference Library is protected by U.S. Pat. No.
//C- | 6,058,214 and patents pending.
//C- |
//C- | This software is subject to, and may be distributed under, the
//C- | GNU General Public License, either Version 2 of the license,
//C- | or (at your option) any later version. The license should have
//C- | accompanied the software or you may obtain a copy of the license
//C- | from the Free Software Foundation at http://www.fsf.org .
//C- |
//C- | The computer code originally released by LizardTech under this
//C- | license and unmodified by other parties is deemed "the LIZARDTECH
//C- | ORIGINAL CODE."  Subject to any third party intellectual property
//C- | claims, LizardTech grants recipient a worldwide, royalty-free, 
//C- | non-exclusive license to make, use, sell, or otherwise dispose of 
//C- | the LIZARDTECH ORIGINAL CODE or of programs derived from the 
//C- | LIZARDTECH ORIGINAL CODE in compliance with the terms of the GNU 
//C- | General Public License.   This grant only confers the right to 
//C- | infringe patent claims underlying the LIZARDTECH ORIGINAL CODE to 
//C- | the extent such infringement is reasonably necessary to enable 
//C- | recipient to make, have made, practice, sell, or otherwise dispose 
//C- | of the LIZARDTECH ORIGINAL CODE (or portions thereof) and not to 
//C- | any greater extent that may be necessary to utilize further 
//C- | modifications or combinations.
//C- |
//C- | The LIZARDTECH ORIGINAL CODE is provided "AS IS" WITHOUT WARRANTY
//C- | OF ANY KIND, EITHER EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED
//C- | TO ANY WARRANTY OF NON-INFRINGEMENT, OR ANY IMPLIED WARRANTY OF
//C- | MERCHANTABILITY OR FITNESS FOR A PARTICULAR PURPOSE.
//C- +------------------------------------------------------------------

#ifdef HAVE_CONFIG_H
# include "config.h"
#endif

#include <stddef.h>
#include <stdlib.h>
#include <stdio.h>
#include <stdarg.h>
#include <string.h>
#include <fcntl.h>
#include <errno.h>

#include "libdjvu/ddjvuapi.h"

#if defined(_WIN32) || defined(__CYGWIN32__)
# include <io.h>
#endif
#if defined(_WIN32) && !defined(__CYGWIN32__)
# include <mbctype.h>
#endif

/* Some day we'll redo i18n right. */
#ifndef i18n
# define i18n(x) (x)
# define I18N(x) (x)
#endif

bool verbose = false;
bool tryhelp = false;
ddjvu_context_t *ctx;
ddjvu_document_t *doc;
ddjvu_job_t *job;

void
progress(int p)
{
  if (verbose)
    {
      int i=0;
      char buffer[52];
      for (; p>0; p-=2)
        buffer[i++]='#';
      for (; i<50;)
        buffer[i++]=' ';
      buffer[i] = 0;
      fprintf(stderr,"\r[%s]",buffer);
    }        
}

void
handle(int wait)
{
  const ddjvu_message_t *msg;
  if (!ctx)
    return;
  if (wait)
    msg = ddjvu_message_wait(ctx);
  while ((msg = ddjvu_message_peek(ctx)))
    {
      switch(msg->m_any.tag)
        {
        case DDJVU_ERROR:
          if (verbose)
            fprintf(stderr,"\n");
          fprintf(stderr,"djvups: %s\n", msg->m_error.message);
          if (msg->m_error.filename)
            fprintf(stderr,"djvups: '%s:%d'\n", 
                    msg->m_error.filename, msg->m_error.lineno);
          if (tryhelp)
            fprintf(stderr,"djvups: %s\n", i18n("Try option --help."));
          exit(10);
        case DDJVU_PROGRESS:
          if (verbose)
            progress(msg->m_progress.percent);
          break;
        default:
          break;
        }
      ddjvu_message_pop(ctx);
    }
}

void 
die(const char *fmt, ...)
{
  /* Handling messages might give a better error message */
  tryhelp = true;
  handle(FALSE);
  /* Print */
  va_list args;
  fprintf(stderr,"ddjvu: ");
  va_start(args, fmt);
  vfprintf(stderr, fmt, args);
  va_end(args);
  fprintf(stderr,"\n");
  /* Terminates */
  exit(10);
}

const char *options[] = {
  I18N("-verbose"),
  I18N("-page=<pagelists>                   (default: print all)"),
  I18N("-format=<ps|eps>                    (default: ps)"),
  I18N("-level=<1|2|3>                      (default: 2)"),
  I18N("-orient=<auto|portrait|landscape>   (default: auto)"),
  I18N("-mode=<color|bw|fore|back>          (default: color)"),
  I18N("-zoom=<auto|25...2400)              (default: auto)"),
  I18N("-color=<yes|no>                     (default: yes)"),
  I18N("-gray                               (same as -color=no)"),
  I18N("-colormatch=<yes|no>                (default: yes)"),
  I18N("-gamma=<0.3...5.0>                  (default: 2.2)"),
  I18N("-copies=<1...999999>                (default: 1)"),
  I18N("-frame=<yes|no>                     (default: no)"),
  I18N("-cropmarks=<yes|no>                 (default: no)"),
#ifdef THIS_THING_DOES_NOT_WORK_WITH_UTF8_STRINGS
  I18N("-text=<yes|no>                      (default: no)"),
#endif
  I18N("-booklet=<no|recto|verso|yes>       (default: no)"),
  I18N("-bookletmax=<n>                     (default: 0)"),
  I18N("-bookletalign=<n>                   (default: 0)"),
  I18N("-bookletfold=<n>[+<m>]              (default: 18+200)"),
  NULL };


void
usage(void)
{
  int i;
#ifdef DJVULIBRE_VERSION
  fprintf(stderr, "DJVUPS --- DjVuLibre-" DJVULIBRE_VERSION "\n");
#endif
  fprintf(stderr, "%s\n",
          i18n("DjVu to PostScript conversion utility\n\n"
               "Usage: djvups [<options>] [<infile.djvu> [<outfile.ps>]]\n"
               "Options:\n  -help"));
  for(i=0; options[i]; i++)
    fprintf(stderr, "  %s\n", i18n(options[i]));
  fprintf(stderr,"\n");
  exit(1);
}

int 
check_option(char *s)
{
  int i;
  for (i=0; options[i]; i++)
    {
      int n = 0;
      const char *p = options[i];
      while (p[n] && p[n]!='=' && p[n]!=' ')
        n += 1;
      if (p[n]=='=' && !strncmp(s, p, n+1))
        return 1;
      if (p[n]!='=' && !strncmp(s, p, n) && !s[n])
        return 1;        
    }
  // compatibility aliases 
  if (!strcmp(s,"-grayscale") ||
      !strncmp(s,"-pages=",7) ||
      !strncmp(s,"-orientation=",13) ||
      !strncmp(s,"-srgb=",6) )
    return 1;
  return 0;
}

int
main(int argc, char **argv)
{

  int i;
  int optc = 0;
  char **optv;
  const char *infile = 0;
  const char *outfile = 0;
  FILE *fout;
#if defined(_WIN32) && !defined(__CYGWIN32__)
  _setmbcp(_MB_CP_OEM);
#endif
  /* Sort options */
  if (! (optv = (char**)malloc(argc*sizeof(char*))))
    die(i18n("Out of memory"));
  for (i=1; i<argc; i++)
    {
      char *s = argv[i];
      if (s[0]=='-' && s[1]=='-')
        s = s+1;
      if (!strcmp(s,"-verbose"))
        verbose = true;
      else if (check_option(s))
        optv[optc++] = s;
      else if (s[0]=='-' && s[1])
        usage();
      else if (s[0] && !infile)
        infile = s;
      else if (s[0] && !outfile)
        outfile = s;
      else
        die(i18n("Incorrect arguments. Try option --help."));
    }
  if (! infile)
    infile = "-";
  if (! outfile)
    outfile = "-";
  /* Open document */
  if (! (ctx = ddjvu_context_create(argv[0])))
    die(i18n("Cannot create djvu context."));
  if (! (doc = ddjvu_document_create_by_filename(ctx, infile, TRUE)))
    die(i18n("Cannot open djvu document '%s'."), infile);
  while (! ddjvu_document_decoding_done(doc))
    handle(TRUE);
  /* Open output file */
  if (! strcmp(outfile,"-")) 
    {
      fout = stdout;
#if defined(__CYGWIN32__)
      setmode(fileno(fout), O_BINARY);
#elif defined(_WIN32)
      _setmode(_fileno(fout), _O_BINARY);
#endif
    } 
  else if (! (fout = fopen(outfile, "wb")))
    die(i18n("Cannot open output file '%s'."), outfile);
  /* Create printing job */
  if (! (job = ddjvu_document_print(doc, fout, optc, optv)))
    die(i18n("Cannot create PostScript conversion job."));
  /* Wait until completion and cleanup */
  while (! ddjvu_job_done(job))
    handle(TRUE);
  if (verbose)
    fprintf(stderr,"\n");
  /* Make sure we get error messages */
  tryhelp = false;
  if (ddjvu_job_error(job))
    handle(FALSE);
  if (ddjvu_job_error(job))
    die(i18n("PostScript conversion job failed."));
  /* Close */
  fclose(fout);
  if (job)
    ddjvu_job_release(job);
  if (doc)
    ddjvu_document_release(doc);
  if (ctx)
    ddjvu_context_release(ctx);
  return 0;
}
